src/runtime/virtcontainers/kata_agent.go

```diff
@@ -1701,29 +1698,33 @@ func (k *kataAgent) handleBlockVolumes(c *Container) ([]*grpc.Storage, error) {
 		// device is detached with detachDevices() for a container.
 		c.devices = append(c.devices, ContainerDevice{ID: id, ContainerPath: m.Destination})
 
-		var vol *grpc.Storage
-
-		device := c.sandbox.devManager.GetDeviceByID(id)
-		if device == nil {
-			k.Logger().WithField("device", id).Error("failed to find device by id")
-			return nil, fmt.Errorf("Failed to find device by id (id=%s)", id)
+		// Create Storage structure
+		vol, err := k.createBlkStorageObject(c, m)
+		if vol == nil || err != nil {
+			return nil, err
 		}
 
-		var err error
-		switch device.DeviceType() {
-		case config.DeviceBlock:
-			vol, err = k.handleDeviceBlockVolume(c, m, device)
-		case config.VhostUserBlk:
-			vol, err = k.handleVhostUserBlkVolume(c, m, device)
-		default:
-			k.Logger().Error("Unknown device type")
-			continue
-		}
+		// The device will be mounted at a unique location within the VM. Mounting
+		// to the container specific location is handled within the OCI spec. Let's ensure that
+		// the storage mount point is unique, and that this is utilized as the source in the OCI
+		// spec.
+		filename := fmt.Sprintf("%s-%s", uuid.Generate().String(), filepath.Base(vol.MountPoint))
+		path := filepath.Join(kataGuestSandboxStorageDir(), filename)
 
-		if vol == nil || err != nil {
-			return nil, err
+		// Update applicable OCI mount source
+		for idx, ociMount := range spec.Mounts {
+			if ociMount.Destination != vol.MountPoint {
+				continue
+			}
+			k.Logger().Debugf("Replacing OCI mount source (%s) with %s", ociMount.Source, path)
```

Comment:
Nit: Might be worth using `%q` to quote the values. But you could also use structure logging. Something like:

```go
k.Logger().WithFields(logrus.Fields{
  "original-source": ociMount.Source,
  "new-source": path,
}).Debug("Replacing OCI mount source")